File: star5/merge_sort.c
/**************************************************************************
 * 連結リスト
 * https://ja.wikipedia.org/wiki/連結リスト
 *
 * マージソート
 * https://ja.wikipedia.org/wiki/マージソート
 * https://programming-place.net/ppp/contents/algorithm/sort/007.html
 **************************************************************************/
#include <stdio.h>
#include <stdlib.h>
#include "linked_list.h"

// 関数プロトタイプ宣言
list_t *merge(list_t *a, list_t *b);
list_t *merge_sort(list_t *list);
// void merge_sort(list_t *list);

int main(int argc, char const *argv[]) {
  // 線形リスト宣言と、一番最初のノード生成
  list_t *list = (list_t *)malloc(sizeof(list_t));

  // このリストにあるノードは一つだけで、
  // 次の要素はまだ空なのでNULLを入れる（番兵）
  list->next  = NULL;
  list->value = 0;

  // 初期値投入
  insert_node(list, 3);
  insert_node(list, 14);
  insert_node(list, 15);
  insert_node(list, 92);
  insert_node(list, 65);
  insert_node(list, 35);
  insert_node(list, 89);
  insert_node(list, 79);
  insert_node(list, 24);
  insert_node(list, 58);

  // リストの表示
  printf("並び替え前\n");
  show_list(list);

  // マージソート実行
  merge_sort(list);

  // リストの表示
  printf("並び替え後\n");
  show_list(list);

  // リストの解放
  release_list(list);

  return 0;
}

// 二つのリストを併合（マージ）する
list_t *merge(list_t *a, list_t *b) {
  list_t result;        // 併合されたリスト
  list_t *w = &result;  // work 変数
                        // これを起点に、a, b 小さい方を順に連結していく

  // リストをマージ
  while (a != NULL && b != NULL) {
    // 昇順ソートのため、小さい要素を先にする。
    // a と b が等しい場合は、先頭を優先すると安定ソートとなる。
    if (a->value <= b->value) {
      w->next = a;        // wの後ろにaを連結する
      w       = w->next;  // aを連結したので、さらに小さい数を連結できるよう
                          // 次のノードに進む
      a       = a->next;  // a の次のノードに進む
    } else {
      w->next = b;
      w       = w->next;
      b       = b->next;
    }
  }

  // 残っている要素を、末尾へ連結
  if (a == NULL) {
    // a が短かったということなので、余っているbのノードを繋ぐ
    w->next = b;
  } else {
    w->next = a;
  }

  // 併合されたリストを返す
  return result.next;
}

// リストを二分割し、再帰的にマージする
list_t *merge_sort(list_t *list) {
  // 動作確認用
  list_t *a;
  list_t *b;
  list_t *w; // work 変数

  // 要素が0又は1であれば終了
  if (list == NULL || list->next == NULL) {
    return list;
  }

  // ２つのリストに分割するため、リストの中心を探す。
  // a はリストの先頭を指し、b はリストの二番目のノードを指す。
  // a が一つ進む間に、b は二つ進むので、
  // b がリストの末尾に辿り着いた時には、
  // a は リストのちょうど中央を指している。
  a = list;
  b = list->next;
  if (b != NULL) {
    b = b->next;
  }
  while (b != NULL) {
    a = a->next;
    b = b->next;
    if (b != NULL) {
      b = b->next;
    }
  }

  // 動作確認用
  // printf("------- a ------- \n");
  // show_list(a);

  // リストを前半と後半に分離する
  // a は リスト中央を指しているので、
  // a->next = NULL とすると list が 前半部分のみになる。
  // この時、後半の始まりが失われないよう、work変数に記憶しておく。
  w = a->next;
  a->next = NULL;

  // 動作確認用
  // printf("------- list ------- \n");
  // show_list(list);

  // 前半と後半のリストを、再帰的にマージする。
  return merge(merge_sort(list), merge_sort(w));
}

File: star5/merge_sort.rb
#==============================================================================
# 連結リスト
# https://ja.wikipedia.org/wiki/連結リスト
#
# マージソート
# https://ja.wikipedia.org/wiki/マージソート
# https://programming-place.net/ppp/contents/algorithm/sort/007.html
#==============================================================================

require "debug"

# 連結リストクラスを読み込む
require "./linked_list_class.rb"

# Listクラスを拡張し、マージソートを実装する
module MergeSort
  refine List do
    # コードが書き易いよう、ノードのアドレスを返すメソッドを定義
    def addr(node)
      self.memory.key(node)
    end

    # コードが書き易いよう、アドレスにあるノードを返すメソッドを定義
    def node(addr)
      self.memory[addr]
    end

    # コードが書き易いよう、ノードに続く後継ノードを返すメソッドを定義
    def succ(node)
      self.memory[node.next]
    end

    # コードが書き易いよう、リストA(自分自身)にノードを連結するメソッドを定義
    def concat(address, other_node)
      # List.newの直後なら、受け取ったノードをそのまま追加して終了
      if self.head.nil?
        self.head = address
        self.memory.store(address, other_node)
        return
      end

      # リスト最後のノードまで進む
      node = self.node(self.head)
      until self.succ(node).nil?
        node = self.succ(node)
      end
      # リスト末尾に連結
      node.next       = address
      self.memory.store(address, other_node)
    end

    # 二つのリストを併合し、併合済のリストwを返すクラスメソッドを定義
    def List.merge(list_a, list_b)
      node_a = list_a.node(list_a.head)
      node_b = list_b.node(list_b.head)
      list_w = List.new

      loop do
        # 大小関係を比較する
        if node_a.value <= node_b.value
          # 値の小さいノードを追加する
          list_w.concat(list_a.addr(node_a), node_a)
          # 次のノードへ進む
          node_a = list_a.succ(node_a)
          if node_a.nil?
            # リストAが尽きたので、リストBの残りを繋ぐ
            until node_b.nil?
              list_w.concat(list_b.addr(node_b), node_b)
              node_b = list_b.succ(node_b)
            end
            # 連結したリストを返す
            return list_w
          end
        else
          # 値の小さいノードを追加する
          list_w.concat(list_b.addr(node_b), node_b)
          # 次のノードへ進む
          node_b = list_b.succ(node_b)
          # debugger
          if node_b.nil?
            # リストBが尽きたので、リストAの残りを繋ぐ
            until node_a.nil?
              list_w.concat(list_a.addr(node_a), node_a)
              node_a = list_a.succ(node_a)
            end
            # 連結したリストを返す
            return list_w
          end
        end
      end
    end

    # リストを二分割し、再帰的にマージする
    def mergesort
      # 長さ0, 若しくは 長さ1のリストなら自分自身を返す
      if self.head.nil? || self.succ(self.node(self.head)).nil?
        return self
      end

      # 長さ2以上なら、前半と後半のリストに分割するため、リストの中心を探す。
      # node_a の 二倍の速度でnode_bは進むため、
      # node_b がリストの末端に達した時、node_a はリストの中央に居る。
      node_a = self.node(self.head)
      node_b = self.succ(node_a)
      until node_b.nil? || self.succ(node_b).nil?
        node_a = self.succ(node_a)
        node_b = self.succ(node_b)
        unless node_b.nil?
          node_b = self.succ(node_b)
        end
      end

      # node_w 以降が後半部分のリストとなる
      node_w = self.succ(node_a)
      # node_a までが前半部分のリストとなる
      node_a.next = nil

      # 前半のリスト
      first_half = List.new
      node_a     = self.node(self.head)
      until node_a.nil?
        first_half.concat(self.addr(node_a), node_a)
        node_a = self.succ(node_a)
      end

      # 後半のリスト
      second_half = List.new
      until node_w.nil?
        second_half.concat(self.addr(node_w), node_w)
        node_w = self.succ(node_w)
      end

      # 前半と後半のリストを再帰的にマージする
      List.merge(first_half.mergesort, second_half.mergesort)
    end
  end
end
using MergeSort

# 線形リスト作成
list = List.new

# 初期値投入
list.insert_node 3
list.insert_node 2
list.insert_node 1
list.insert_node 4
list.insert_node 8
list.insert_node 7
list.insert_node 5
list.insert_node 6
list.insert_node 9

puts "--- 並び替え前 ---"
pp list
list.show
# #<List:0x0000000104e042a0
#  @head=100,
#  @memory=
#   {100=>#<struct List::Node next=120, value=3>,
#    120=>#<struct List::Node next=140, value=2>,
#    140=>#<struct List::Node next=160, value=1>,
#    160=>#<struct List::Node next=180, value=4>,
#    180=>#<struct List::Node next=200, value=8>,
#    200=>#<struct List::Node next=220, value=7>,
#    220=>#<struct List::Node next=240, value=5>,
#    240=>#<struct List::Node next=260, value=6>,
#    260=>#<struct List::Node next=nil, value=9>}>
# 3 2 1 4 8 7 5 6 9

# マージソート実行
list = list.mergesort

puts "--- 並び替え後 ---"
pp list
list.show
# #<List:0x0000000104c3f398
#  @head=140,
#  @memory=
#   {140=>#<struct List::Node next=120, value=1>,
#    120=>#<struct List::Node next=100, value=2>,
#    100=>#<struct List::Node next=160, value=3>,
#    160=>#<struct List::Node next=220, value=4>,
#    220=>#<struct List::Node next=240, value=5>,
#    240=>#<struct List::Node next=200, value=6>,
#    200=>#<struct List::Node next=180, value=7>,
#    180=>#<struct List::Node next=260, value=8>,
#    260=>#<struct List::Node next=nil, value=9>}>
# 1 2 3 4 5 6 7 8 9
